pkg/util/tas/tas_assignment_performance_test.go

```diff
@@ -0,0 +1,279 @@
+			)
+		}
+		return res
+	}
+}
+
+type regionAndIPBasedNamingConfig struct {
+	fixedPrefixAndSuffixLength int
+	regions                    int
+	regionIDLength             int
+}
+
+func regionAndIPBasedNaming(config regionAndIPBasedNamingConfig) namingScheme {
+	return func(nodes int) []string {
+		res := make([]string, nodes)
+		nodeIPs := consecutiveIPs(nodes)
+		regionIDs := randomHexIDs(config.regions, config.regionIDLength)
+		for i := range nodes {
+			res[i] = fmt.Sprintf("%s-%s-%s-%s",
+				fixedID(config.fixedPrefixAndSuffixLength),
+				regionIDs[i%config.regions],
+				nodeIPs[i],
+				fixedID(config.fixedPrefixAndSuffixLength),
+			)
+		}
+		return res
+	}
+}
+
+type nodeBasedNamingConfig struct {
+	fixedPrefixAndSuffixLength int
+	nodeIDLength               int
+}
+
+func nodeBasedNaming(config nodeBasedNamingConfig) namingScheme {
+	return func(nodes int) []string {
+		res := make([]string, nodes)
+		nodeIDs := randomHexIDs(nodes, config.nodeIDLength)
+		for i := range nodes {
+			res[i] = fmt.Sprintf("%s-%s-%s",
+				fixedID(config.fixedPrefixAndSuffixLength),
+				nodeIDs[i],
+				fixedID(config.fixedPrefixAndSuffixLength),
+			)
+		}
+		return res
+	}
+}
+
+func internalSinglePodsOn(nodes []string) *TopologyAssignment {
+	res := &TopologyAssignment{
+		Levels:  []string{corev1.LabelHostname},
+		Domains: make([]TopologyDomainAssignment, len(nodes)),
+	}
+	for i, n := range nodes {
+		res.Domains[i] = TopologyDomainAssignment{
+			Values: []string{n},
+			Count:  1,
+		}
+	}
+	return res
+}
+
+func jsonStr(v any) string {
+	bytes, err := json.Marshal(v)
+	if err != nil {
+		panic(err)
+	}
+	return string(bytes)
+}
+
+func isTooLarge(ta *kueue.TopologyAssignment) bool {
+	return len(jsonStr(ta)) > 1_500_000
+}
+
+func approxMaxNodesFor(naming namingScheme) int {
+	step := 1_000 // We search with a reduced resolution, to speed up the test
+	ceiling := 300_000
+	nodeNames := naming(ceiling)
+	found := sort.Search(ceiling/step, func(n int) bool {
+		// Here we rely on "well-prefixing"; see the comment on "nodeNaming".
+		return isTooLarge(V1Beta2From(internalSinglePodsOn(nodeNames[:n*step])))
+	}) - 1
+	return found * step
+}
+
+type performanceTestCase struct {
+	name   string
+	naming namingScheme
+}
+
+var performanceTestCases = []performanceTestCase{
+	{
+		name: "pool-and-node-based naming (100 node pools)",
+		naming: poolAndNodeBasedNaming(poolAndNodeBasedNamingConfig{
+			pools:        100, // happens in practice, at least in GKE
+			nodeIDLength: 6,   // reached in AKS
+
+			// reachable in AKS (<pool-name>-<8-char-id>-vmss, then let <pool-name> have 8 chars)
+			poolIDLength: 22,
+
+			fixedPrefixAndSuffixLength: 20,
+		}),
+	},
+	{
+		name: "pool-and-node-based naming (1 node pool)",
+		naming: poolAndNodeBasedNaming(poolAndNodeBasedNamingConfig{
+			pools:                      1,
+			nodeIDLength:               6,
+			poolIDLength:               22,
+			fixedPrefixAndSuffixLength: 20,
+		}),
```

Comment:
Do we know about any vendor that could fit thousands of into a single nodepool (or equivalent concept)

Reply:
In GKE, the maximum node count in a node pool seems to be 4000.
I failed to find a vendor (with "pools" reflected in node names) surpassing that.

So indeed, as long as we care about 40k+ nodes, it seems 10+ node pools is a must.
To have more realistic scenarios, I bumped 1 node pool -> 10.
And BTW I also bumped 100 NPs -> 1000, as this always sometimes exists in GKE.